File: qt_ui/main_window/window.cpp
// Main Window UI Shell Implementation

#include "window.hpp"
#include "controller.hpp"

#include <QVBoxLayout>
#include <QHBoxLayout>
#include <QGroupBox>
#include <QWidget>

MainWindow::MainWindow(QWidget* parent)
    : QMainWindow(parent)
{
    setWindowTitle("Video/Audio Sync & Merge");
    setGeometry(100, 100, 1000, 600);

    buildUi();

    // Create controller and wire up signals
    m_controller = std::make_unique<MainController>(this);
    m_controller->applyConfigToUi();

    // Connect signals to controller slots
    connect(m_optionsBtn, &QPushButton::clicked,
            m_controller.get(), &MainController::openOptionsDialog);
    connect(m_queueJobsBtn, &QPushButton::clicked,
            m_controller.get(), &MainController::openJobQueue);
    connect(m_analyzeBtn, &QPushButton::clicked,
            m_controller.get(), &MainController::startAnalyzeOnly);
}

MainWindow::~MainWindow() = default;

void MainWindow::buildUi()
{
    auto* central = new QWidget(this);
    setCentralWidget(central);
    auto* mainLayout = new QVBoxLayout(central);

    // =========================================================================
    // Top Row - Settings button
    // =========================================================================
    auto* topRow = new QHBoxLayout();
    m_optionsBtn = new QPushButton("Settings...");
    topRow->addWidget(m_optionsBtn);
    topRow->addStretch();
    mainLayout->addLayout(topRow);

    // =========================================================================
    // Main Workflow Group
    // =========================================================================
    auto* actionsGroup = new QGroupBox("Main Workflow");
    auto* actionsLayout = new QVBoxLayout(actionsGroup);

    m_queueJobsBtn = new QPushButton("Open Job Queue for Merging...");
    m_queueJobsBtn->setStyleSheet("font-size: 14px; padding: 5px;");
    actionsLayout->addWidget(m_queueJobsBtn);

    m_archiveLogsCheck = new QCheckBox("Archive logs to a zip file on batch completion");
    actionsLayout->addWidget(m_archiveLogsCheck);

    mainLayout->addWidget(actionsGroup);

    // =========================================================================
    // Quick Analysis Group
    // =========================================================================
    auto* analysisGroup = new QGroupBox("Quick Analysis (Analyze Only)");
    auto* analysisLayout = new QVBoxLayout(analysisGroup);

    // Create input fields
    m_refInput = new QLineEdit();
    m_secInput = new QLineEdit();
    m_terInput = new QLineEdit();

    // Add file input rows
    analysisLayout->addLayout(createFileInput(
        "Source 1 (Reference):", m_refInput,
        [this]() { m_controller->browseForPath(m_refInput, "Select Reference File or Directory"); }
    ));
    analysisLayout->addLayout(createFileInput(
        "Source 2:", m_secInput,
        [this]() { m_controller->browseForPath(m_secInput, "Select Secondary File or Directory"); }
    ));
    analysisLayout->addLayout(createFileInput(
        "Source 3:", m_terInput,
        [this]() { m_controller->browseForPath(m_terInput, "Select Tertiary File or Directory"); }
    ));

    // Analyze button (right-aligned)
    auto* analyzeBtnLayout = new QHBoxLayout();
    analyzeBtnLayout->addStretch();
    m_analyzeBtn = new QPushButton("Analyze Only");
    analyzeBtnLayout->addWidget(m_analyzeBtn);
    analysisLayout->addLayout(analyzeBtnLayout);

    mainLayout->addWidget(analysisGroup);

    // =========================================================================
    // Status Row
    // =========================================================================
    auto* statusLayout = new QHBoxLayout();
    statusLayout->addWidget(new QLabel("Status:"));
    m_statusLabel = new QLabel("Ready");
    statusLayout->addWidget(m_statusLabel, 1);
    m_progressBar = new QProgressBar();
    m_progressBar->setRange(0, 100);
    m_progressBar->setValue(0);
    m_progressBar->setTextVisible(true);
    statusLayout->addWidget(m_progressBar);
    mainLayout->addLayout(statusLayout);

    // =========================================================================
    // Latest Job Results Group
    // =========================================================================
    auto* resultsGroup = new QGroupBox("Latest Job Results");
    auto* resultsLayout = new QHBoxLayout(resultsGroup);

    // Create delay labels for Source 2, 3, 4
    for (int i = 2; i <= 4; ++i) {
        resultsLayout->addWidget(new QLabel(QString("Source %1 Delay:").arg(i)));
        auto* delayLabel = new QLabel(QString::fromUtf8("\u2014")); // em-dash
        m_delayLabels.push_back(delayLabel);
        resultsLayout->addWidget(delayLabel);
        resultsLayout->addSpacing(20);
    }
    resultsLayout->addStretch();

    mainLayout->addWidget(resultsGroup);

    // =========================================================================
    // Log Group
    // =========================================================================
    auto* logGroup = new QGroupBox("Log");
    auto* logLayout = new QVBoxLayout(logGroup);
    m_logOutput = new QTextEdit();
    m_logOutput->setReadOnly(true);
    m_logOutput->setFontFamily("monospace");
    logLayout->addWidget(m_logOutput);
    mainLayout->addWidget(logGroup);
}

QLayout* MainWindow::createFileInput(const QString& label, QLineEdit* input,
                                      const std::function<void()>& browseCallback)
{
    auto* layout = new QHBoxLayout();
    auto* labelWidget = new QLabel(label);
    labelWidget->setMinimumWidth(140);
    layout->addWidget(labelWidget);
    layout->addWidget(input, 1);

    auto* browseBtn = new QPushButton("Browse...");
    connect(browseBtn, &QPushButton::clicked, this, browseCallback);
    layout->addWidget(browseBtn);

    return layout;
}

File: Reference Only original/vsg_qt/main_window/window.py
# vsg_qt/main_window/window.py
from __future__ import annotations

from PySide6.QtCore import Qt
from PySide6.QtWidgets import (
    QCheckBox,
    QGroupBox,
    QHBoxLayout,
    QLabel,
    QLineEdit,
    QMainWindow,
    QProgressBar,
    QPushButton,
    QTextEdit,
    QVBoxLayout,
    QWidget,
)

from vsg_core.config import AppConfig

from .controller import MainController


class MainWindow(QMainWindow):
    """Slim UI shell: builds widgets & delegates logic to MainController."""

    def __init__(self):
        super().__init__()
        self.setWindowTitle("Video/Audio Sync & Merge - PySide6 Edition")
        self.setGeometry(100, 100, 1000, 600)

        self.config = AppConfig()
        self.controller = MainController(self)

        self._build_ui()
        self.controller.apply_config_to_ui()

    def _build_ui(self):
        # Quick Analysis Inputs
        self.ref_input = QLineEdit()
        self.sec_input = QLineEdit()
        self.ter_input = QLineEdit()

        # Log and Status
        self.log_output = QTextEdit()
        self.log_output.setReadOnly(True)
        self.log_output.setFontFamily("monospace")
        self.progress_bar = QProgressBar()
        self.progress_bar.setRange(0, 100)
        self.progress_bar.setValue(0)
        self.progress_bar.setTextVisible(True)
        self.status_label = QLabel("Ready")

        # Delay Labels
        self.delay_labels: list[QLabel] = []

        # Other Controls
        self.archive_logs_check = QCheckBox(
            "Archive logs to a zip file on batch completion"
        )

        central = QWidget()
        self.setCentralWidget(central)
        main_layout = QVBoxLayout(central)

        top_row = QHBoxLayout()
        self.options_btn = QPushButton("Settings…")
        top_row.addWidget(self.options_btn)
        top_row.addStretch()
        main_layout.addLayout(top_row)

        actions_group = QGroupBox("Main Workflow")
        actions_layout = QVBoxLayout(actions_group)
        self.queue_jobs_btn = QPushButton("Open Job Queue for Merging...")
        self.queue_jobs_btn.setStyleSheet("font-size: 14px; padding: 5px;")
        actions_layout.addWidget(self.queue_jobs_btn)
        actions_layout.addWidget(self.archive_logs_check)
        main_layout.addWidget(actions_group)

        analysis_group = QGroupBox("Quick Analysis (Analyze Only)")
        analysis_layout = QVBoxLayout(analysis_group)
        analysis_layout.addLayout(
            self._create_file_input(
                "Source 1 (Reference):",
                self.ref_input,
                lambda: self.controller.browse_for_path(
                    self.ref_input, "Select Reference File or Directory"
                ),
            )
        )
        analysis_layout.addLayout(
            self._create_file_input(
                "Source 2:",
                self.sec_input,
                lambda: self.controller.browse_for_path(
                    self.sec_input, "Select Secondary File or Directory"
                ),
            )
        )
        analysis_layout.addLayout(
            self._create_file_input(
                "Source 3:",
                self.ter_input,
                lambda: self.controller.browse_for_path(
                    self.ter_input, "Select Tertiary File or Directory"
                ),
            )
        )
        analyze_btn = QPushButton("Analyze Only")
        analysis_layout.addWidget(analyze_btn, 0, Qt.AlignRight)
        main_layout.addWidget(analysis_group)

        # Connect signals
        self.options_btn.clicked.connect(self.controller.open_options_dialog)
        analyze_btn.clicked.connect(self.controller.start_batch_analyze_only)
        self.queue_jobs_btn.clicked.connect(self.controller.open_job_queue)

        status_layout = QHBoxLayout()
        status_layout.addWidget(QLabel("Status:"))
        status_layout.addWidget(self.status_label, 1)
        status_layout.addWidget(self.progress_bar)
        main_layout.addLayout(status_layout)

        results_group = QGroupBox("Latest Job Results")
        results_layout = QHBoxLayout(results_group)
        # DYNAMIC FIX: Create a few labels for delays
        for i in range(2, 5):  # Create labels for Source 2, 3, 4
            results_layout.addWidget(QLabel(f"Source {i} Delay:"))
            delay_label = QLabel("—")
            self.delay_labels.append(delay_label)
            results_layout.addWidget(delay_label)
            results_layout.addSpacing(20)
        # Store references to the first two for backward compatibility if needed elsewhere
        self.sec_delay_label = self.delay_labels[0]
        self.ter_delay_label = self.delay_labels[1]
        results_layout.addStretch()
        main_layout.addWidget(results_group)

        log_group = QGroupBox("Log")
        log_layout = QVBoxLayout(log_group)
        log_layout.addWidget(self.log_output)
        main_layout.addWidget(log_group)

    def _create_file_input(self, label_text: str, line_edit: QLineEdit, on_browse):
        layout = QHBoxLayout()
        layout.addWidget(QLabel(label_text), 1)
        layout.addWidget(line_edit, 8)
        browse_btn = QPushButton("Browse…")
        browse_btn.clicked.connect(on_browse)
        layout.addWidget(browse_btn, 1)
        return layout

    def closeEvent(self, event):
        self.controller.on_close()
        super().closeEvent(event)
